File: 2024/16/a.py
from itertools import groupby, product
from pathlib import Path

p1 = p2 = 0

INPUT_FILE = Path(__file__).parent / "input"

contents = INPUT_FILE.read_text().rstrip()
lines = contents.splitlines()
sections = [
    list(section)
    for has_content, section in groupby(lines, bool)
    if has_content
]

grid = {}
ipos = None
epos = None
for y, line in enumerate(lines):
    for x, c in enumerate(line):
        grid[x, y] = c
        if c == 'S':
            ipos = x, y
        if c == 'E':
            epos = x, y

w, h = len(line), len(lines)


directions = (0, -1), (1, 0), (0, 1), (-1, 0)

def allowed(direction):
    i = directions.index(direction)

    return directions[i-1], directions[(i+1) % len(directions)]

def inverse(direction):
    return directions[directions.index(direction)-2]

from bisect import insort
from collections import deque, defaultdict

RIGHT = (1, 0)

remaining = deque([(0, ipos, RIGHT, {ipos})])

from math import inf

seen = defaultdict(lambda: inf)
istate = ipos, RIGHT
seen[istate] = 0


best = None

in_best = set()

while remaining:
    score, pos, direction, history = remaining.popleft()
    x, y = pos

    if best is not None:
        if score > best:
            continue

    if pos == epos:
        best = score
        in_best.update(history)
        continue

    dx, dy = direction

    npos = nx, ny = x+dx, y+dy
    if grid.get(npos) != '#':
        state = npos, direction
        nscore = score + 1

        if seen[state] >= nscore:
            seen[state] = nscore
            insort(remaining, (nscore, npos, direction, history | {npos}))

    nscore = score + 1000
    for ndirection in allowed(direction):
        state = pos, ndirection
        if seen[pos, inverse(ndirection)] != inf:
            continue

        if grid.get((pos[0] + ndirection[0], pos[1] + ndirection[1])) != '.':
            continue

        if seen[state] >= nscore:
            insort(remaining, (nscore, pos, ndirection, history))

print(best)
print(len(in_best))

File: 2024/16/a.nim
#[
  https://learnxinyminutes.com/nim/
  https://nim-lang.org/docs/manual.html
  https://nimbyexample.com/

  run with
    nim r a.nim
  alternatively
    nim r -d:release a.nim
]#

import std/enumerate
import std/heapqueue
import std/sequtils
import std/sets
import std/tables

type
  Position     = tuple[x: int, y: int]
  Displacement = tuple[x: int, y: int]
  Direction = enum north, east, south, west
  Pose = tuple[pos: Position, dir: Direction]
  State = tuple[score: int, p: Pose, history: HashSet[Position]]

func left(dir: Direction): Direction =
  return (if dir == north: west else: dir.pred)

func right(dir: Direction): Direction =
  return (if dir == west: north else: dir.succ)

func displacement(dir: Direction): Displacement =
  case dir
  of north: return ( 0, -1)
  of east:  return ( 1,  0)
  of south: return ( 0,  1)
  of west:  return (-1,  0)

func `+`(pos: Position, d: Displacement): Position =
  return (pos.x + d.x, pos.y + d.y)

func `+`(pos: Position, d: Direction): Position =
  return pos + d.displacement

func `<`(a: State, b: State): bool =
  return a.score < b.score

func ahead(p: Pose): Pose =
  return (p.pos + p.dir, p.dir)

func left(p: Pose): Pose =
  return (p.pos, p.dir.left).ahead

func right(p: Pose): Pose =
  return (p.pos, p.dir.right).ahead

func next(s: State): seq[State] =
  return @[
    (s.score + 1, s.p.ahead, s.history + toHashSet([s.p.ahead.pos])),
    (s.score + 1001, s.p.right, s.history + toHashSet([s.p.right.pos])),
    (s.score + 1001, s.p.left,  s.history + toHashSet([s.p.left.pos ]))
  ]

var
  grid = initTable[Position, char]()
  ipos: Position
  epos: Position

for (y, line) in enumerate("input".lines):
  for (x, c) in enumerate(line):
    var pos = (x, y)

    grid[pos] = c
    if c == 'S':
      ipos = pos
    if c == 'E':
      epos = pos

var
  ip = (ipos, east)
  istate = (0, ip, toHashSet[Position]([ipos]))
  in_best = initHashSet[Position]()
  remaining = toHeapQueue[State]([istate])
  seen = toTable[Pose, int]({ip: 0})
  best = -1

proc shouldCheck(s: State): bool =
  if grid[s.p.pos] == '#':
    return false

  if s.p in seen and s.score > seen[s.p]:
    return false

  return true

while remaining.len > 0:
  let curr = remaining.pop

  if best != -1 and curr.score > best:
    continue

  if curr.p.pos == epos:
    in_best = in_best.union(curr.history)
    if best == -1:
      echo curr.score
      best = curr.score

    continue

  for s in curr.next.filter(shouldCheck):
    seen[s.p] = s.score
    remaining.push(s)

echo in_best.len
